File: sections/41_euler_tour_lca_binary_lifting/problems/c_kth_node_path/solution.py
import sys
def main():
    data=list(map(int,sys.stdin.buffer.read().split()))
    if not data: return
    n,q=data[0],data[1]; idx=2; LOG=max(1,(n+1).bit_length())
    up=[[0]*(n+1) for _ in range(LOG)]; depth=[0]*(n+1)
    for v in range(2,n+1):
        p=data[idx]; idx+=1; up[0][v]=p; depth[v]=depth[p]+1
    for k in range(1,LOG):
        for v in range(1,n+1): up[k][v]=up[k-1][up[k-1][v]]
    def lift(v,d):
        for k in range(LOG):
            if d>>k & 1: v=up[k][v]
        return v
    def lca(a,b):
        if depth[a]<depth[b]: a,b=b,a
        a=lift(a,depth[a]-depth[b])
        if a==b: return a
        for k in range(LOG-1,-1,-1):
            if up[k][a]!=up[k][b]: a=up[k][a]; b=up[k][b]
        return up[0][a]
    out=[]
    for _ in range(q):
        u,v,k=data[idx],data[idx+1],data[idx+2]; idx+=3; c=lca(u,v); left=depth[u]-depth[c]+1; total=depth[u]+depth[v]-2*depth[c]+1
        if k<1 or k>total: out.append("-1")
        elif k<=left: out.append(str(lift(u,k-1)))
        else: out.append(str(lift(v,total-k)))
    print("\n".join(out))
if __name__=="__main__": main()

File: sections/41_euler_tour_lca_binary_lifting/problems/c_kth_node_path/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, q;
    if (!(cin >> n >> q)) return 0;

    int LOG = 1;
    while ((1 << LOG) <= n + 1) LOG++;

    vector<vector<int>> up(LOG, vector<int>(n + 1));
    vector<int> depth(n + 1);

    for (int v = 2; v <= n; v++) {
        cin >> up[0][v];
        depth[v] = depth[up[0][v]] + 1;
    }
    for (int k = 1; k < LOG; k++)
        for (int v = 1; v <= n; v++)
            up[k][v] = up[k - 1][up[k - 1][v]];

    auto lift = [&](int v, int d) {
        for (int k = 0; k < LOG; k++)
            if (d & (1 << k)) v = up[k][v];
        return v;
    };

    auto lca = [&](int a, int b) {
        if (depth[a] < depth[b]) swap(a, b);
        a = lift(a, depth[a] - depth[b]);
        if (a == b) return a;
        for (int k = LOG - 1; k >= 0; k--)
            if (up[k][a] != up[k][b]) {
                a = up[k][a];
                b = up[k][b];
            }
        return up[0][a];
    };

    while (q--) {
        int u, v, k;
        cin >> u >> v >> k;
        int c = lca(u, v);
        int left = depth[u] - depth[c] + 1;
        int total = depth[u] + depth[v] - 2 * depth[c] + 1;
        if (k < 1 || k > total)
            cout << -1 << '\n';
        else if (k <= left)
            cout << lift(u, k - 1) << '\n';
        else
            cout << lift(v, total - k) << '\n';
    }
}
